### PR TITLE
JIT: Fold unreachable cases for switch in importer.

### LINKED ISSUE #65368: JIT: fold unreachable cases for switch
```csharp
static void Foo() => Bar(5);

static void Bar(int i)
{
    switch (i)
    {
        case 0:
            Console.WriteLine();
            break;
        case 1:
            Console.WriteLine();
            break;
        case 2:
            Console.WriteLine();
            break;
        case 3:
            Console.WriteLine();
            break;
        case 4:
            Console.WriteLine();
            break;
        case 5:
            // no-op
            break;
    }
}
```

After `Bar` is inlined into `Foo` we should get:
```asm
; Method Program:Foo()
       C3                   ret      
; Total bytes of code: 1
```
and we do lower it to exactly this already! The problem that it seems the importer is not able to do it early and materializes all trees:
image

so these calls (unreachable cases) live till some late phase (`PHASE Optimize layout` to be precise). If we fix it we might improve throughput for some methods.

It's actually quite important for this method:
```csharp
    static bool Equals(ReadOnlySpan<char> a, ReadOnlySpan<char> b) => 
        a.Equals(b, StringComparison.Ordinal);
```
Here we pass `StringComparison.Ordinal` (constant) to a switch inside Equals, see here https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs#L46-L67

so JIT ends up importing everything and only then does dead code elimination. it leads to this codegen: https://gist.github.com/EgorBo/7117c2fc53892cf17ade0f11ddd67240 (the final codegen is more or less fine but just take a look at the amount of temp locals)

category:cq
theme:basic-cq

### PR DESCRIPTION
**Pull request by JulieLeeMSFT**
This PR folds unreachable cases in Switch early in `Import` phase when switch value is GT_CNT_INT. 
Fixes https://github.com/dotnet/runtime/issues/65368.

Below trees show the optimization example for the below scenario:
```
static void Foo() => Bar(3);

static void Bar(int i)
{
    switch (i)
    {
        case 0:
            Console.WriteLine();
            break;
        case 1:
            Console.WriteLine();
            break;
        case 2:
            Console.WriteLine();
            break;
        case 3:
            Console.WriteLine();
            break;
        case 4:
            Console.WriteLine();
            break;
        case 5:
            // no-op
            break;
    }
}
```

- Trees before Switch folding:
```
*************** Inline @[000003] Finishing PHASE Profile incorporation
Trees after Profile incorporation

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB03 [0002]  1                             1       [000..01E)-> BB05,BB06,BB07,BB08,BB09,BB10,BB04[def] (switch)                     
BB04 [0003]  1       BB03                  1       [01E..01F)        (return)                     
BB05 [0004]  1       BB03                  1       [01F..025)        (return)                     
BB06 [0005]  1       BB03                  1       [025..02B)        (return)                     
BB07 [0006]  1       BB03                  1       [02B..031)        (return)                     
BB08 [0007]  1       BB03                  1       [031..037)        (return)                     
BB09 [0008]  1       BB03                  1       [037..03C)                                     
BB10 [0009]  2       BB03,BB09             1       [03C..03D)        (return)                     
-----------------------------------------------------------------------------------------------------------------------------------------
```

- After Importation, BB03 always jumps to BB08 and other basic blocks are folded away:
```
*************** Inline @[000003] Finishing PHASE Importation
Trees after Importation

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB03 [0002]  1                             1       [000..01E)-> BB08 (always)                     i 
BB04 [0003]  0                             1       [01E..01F)        (return)                     
BB05 [0004]  0                             1       [01F..025)        (return)                     
BB06 [0005]  0                             1       [025..02B)        (return)                     
BB07 [0006]  0                             1       [02B..031)        (return)                     
BB08 [0007]  1       BB03                  1       [031..037)        (return)                     i 
BB09 [0008]  0                             1       [037..03C)                                     
BB10 [0009]  1       BB09                  1       [03C..03D)        (return)                     
-----------------------------------------------------------------------------------------------------------------------------------------
```

```
*************** Before renumbering the basic blocks

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB03 [0002]  1                             1       [000..01E)-> BB08 (always)                     i 
BB08 [0007]  1       BB03                  1       [031..037)        (return)                     i 
-----------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table is empty
Renumber BB08 to BB04
```

```
*************** Inline @[000003] Finishing PHASE Post-import
Trees after Post-import

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB03 [0002]  1                             1       [000..01E)-> BB04 (always)                     i 
BB04 [0007]  1       BB03                  1       [031..037)        (return)                     i 
-----------------------------------------------------------------------------------------------------------------------------------------

------------ BB03 [000..01E) -> BB04 (always), preds={} succs={BB04}

***** BB03
STMT00003 ( 0x000[E-] ... ??? ) <- INL01 @ 0x000[E-] <- INLRT @ 0x000[E-]
               [000006] -----------                         *  NOP       void  

------------ BB04 [031..037) (return), preds={BB03} succs={}

***** BB04
STMT00004 ( 0x031[E-] ... ??? ) <- INL01 @ 0x000[E-] <- INLRT @ 0x000[E-]
               [000007] --C-G------                         *  CALL      void   System.Console:WriteLine()
```

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@JulieLeeMSFT is this ready for review?

**Comment by AndyAyersMS** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=188428&view=ms.vss-build-web.run-extensions-tab) show some unexpected size increases, eg
```
;; x64 windows libraries_tests.pmi.windows.x64.checked.mch

Top method regressions (percentages):
         800 (37.54 % of base) : 175521.dasm - System.Runtime.Serialization.Xml.XsdDataContractExporterTests.ExporterApiTests+<>c:<Export_MemberData>b__7_7(System.String,System.Xml.Schema.XmlSchemaSet):this
         630 (37.21 % of base) : 57285.dasm - System.Text.Json.Serialization.Tests.PolymorphicTests:<NestedObjectAsRootObject>g__Verify|6_0(System.String)
         420 (31.37 % of base) : 56898.dasm - System.Text.Json.Serialization.Tests.NullTests:NullReferences()
```
Can you investigate?

Also some diffs in minopts which we probably don't want.

**Comment by JulieLeeMSFT** [comment]
Yes, I am working on it.

> Can you investigate?

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -7993,13 +7993,83 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                           info.compFullName);
+                }
+#endif
                 /* Pop the switch value off the stack */
                 op1 = impPopStack().val;
                 assertImp(genActualTypeIsIntOrI(op1->TypeGet()));
 
-                /* We can create a switch node */
+                // Fold Switch for GT_CNS_INT
+                if ((op1->gtOper == GT_CNS_INT) && !compIsForImportOnly())
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:8008]:
```suggestion
                if (opts.OptimizationEnabled() && (op1->gtOper == GT_CNS_INT) && !compIsForImportOnly())
```
We don't want to do this unless we're optimizing the method.

This should fix the diffs you are seeing with minopts.

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -7993,13 +7993,83 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                    printf("\n---- CEE_SWITCH start in " FMT_BB " (PC=%03u) of '%s'", block->bbNum, block->bbCodeOffs,
+                           info.compFullName);
+                }
+#endif
                 /* Pop the switch value off the stack */
                 op1 = impPopStack().val;
                 assertImp(genActualTypeIsIntOrI(op1->TypeGet()));
 
-                /* We can create a switch node */
+                // Fold Switch for GT_CNS_INT
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:8007]: You should try constant-folding the `op1` here, to increase the number of times we see constants.

```C++
                op1 = gtFoldExpr(impPopStack().val);
```

(replace line 8004 above)

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -7993,13 +7993,83 @@ void Compiler::impImportBlockCode(BasicBlock* block)
...
+                    {
+                        printf("\nSwitch folded at " FMT_BB "\n", block->bbNum);
+                        printf(FMT_BB " becomes a %s", block->bbNum,
+                               block->bbJumpKind == BBJ_ALWAYS ? "BBJ_ALWAYS" : "BBJ_NONE");
+                        if (block->bbJumpKind == BBJ_ALWAYS)
+                        {
+                            printf(" to " FMT_BB, block->bbJumpDest->bbNum);
+                        }
+                        printf("\n");
+                        fgDispBasicBlocks(true);
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importer.cpp:8060]: We probably don't need to show all BBs here?

**Comment by JulieLeeMSFT** [comment]
[Diff](https://dev.azure.com/dnceng-public/public/_build/results?buildId=203911&view=ms.vss-build-web.run-extensions-tab)
```
Size: -6216/+10973 
Throughput: up to -0.06% (improved)
 ```

image
image

**Comment by JulieLeeMSFT** [comment]
Size regressions are good diffs or incidental changes. So, all look good. Some examples:
- Case 1: In diff, some inlines in Switch arm got folded away. And, later there are inlines in the diff that were not done in the base becasue the diff compiler did not do those extra inlines, it has fewer local vars, and that made a later inline more attractive. So, this is good diffs.
image
- Case 2: 2 jumps are removed. So, it is a good diff.
image
- Case 3: Incidental size increase due to removed "SHORT".
image

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (JulieLeeMSFT) [reply src/coreclr/jit/importer.cpp:8007]: Done.

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (JulieLeeMSFT) [reply src/coreclr/jit/importer.cpp]: Done.

File: src/coreclr/jit/importer.cpp (hunk 3 above)
Comment (JulieLeeMSFT) [reply src/coreclr/jit/importer.cpp]: Removed.

**Comment by JulieLeeMSFT** [comment]
@AndyAyersMS, it is ready for review.
CC @dotnet/jit-contrib.
All tests passed in the commit f9b19b766386254d924b484ee224d45cd3123737.

**Review by AndyAyersMS — APPROVED** [review]
Looks good!

**Pull request merged**